### ISSUE TITLE
havoc assignment does not respect newtype or subset constraints

### ISSUE THREAD

**Issue by davidcok** [dafny-lang/dafny#3742]
### Dafny version

4.0.0

### Code to produce this issue

```dafny
newtype n10 = x: int | 0 <= x < 10

method b(x: n10) {
  assert 0 <= (x as int) < 10;
}

method bd() {
  var x: n10 := *;
  assert 0 <= (x as int) < 10;
}
```

### Command to run and resulting output

```code
dafny verify
```

### What happened?

The first assert verifies, but not the second.
Same behavior for a subset type and a newtype.

### What type of operating system are you experiencing the problem on?

Mac

**Comment by RustanLeino** [comment]
Note, this bug appears with the strict definite-assignment mode (default with the new CLI, and `/definiteAssignment:4` with the legacy CLI).

**Comment by RustanLeino** [comment]
You're right that the problem is related to the `havoc` and the `where` clause. In Boogie, a `where` clause is a condition that gets assumed anytime that Boogie is allowed to set the variable to an arbitrary value. This happens at the very beginning of a procedure body (both for in-parameters and local variables), on return from a call (for the out-parameters), and in `havoc` statements. Thus, the `havoc x#0;` above lets Boogie set `x#0` to any value that satisfies the condition

    defass#x#0 ==> LitInt(0) <= x#0 && x#0 < 10

However, at the time of this `havoc`, the value of `defass#x#0` may be `false`, in which case the new value of `x#0` can be anything. Immediately following the `havoc` in the code above, `defass#x#0` is set to `true`, but that's too late to affect the `havoc` on the previous line.

So, what needs to happen is that the assignment `defass#x#0 := true;` should be done just before, not after, the `havoc x#0;`. (But it should still happen _after_ the well-formedness checking of the RHSs and LHSs of the Dafny assignment.)

Here is another example, which tests that the well-formedness of the RHSs are checked before the `defass` variables are updated:

``` dafny
method R() returns (r: nat) {
  var a: nat;
  a, r := *, a;

  var b: nat;
  r, b := b, *;

  var c: nat;
  c := *;

  assert 0 <= a && 0 <= b && 0 <= c && 0 <= r;
}
```
